### PR TITLE
[MDEPLOY-206] Support parallel deployment for deployAtEnd

### PR DESCRIPTION
**Pull request by slawekjaranowski**
Each whole module will be deployed by resolver in one thread. 
We have parallel deployment on module level.

---

Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MDEPLOY) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MDEPLOY-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MDEPLOY-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [x] You have run the integration tests successfully (`mvn -Prun-its clean verify`).

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Comment by gnodet** [comment]
@slawekjaranowski the code for multi-threaded deploy should work if the number of threads is set to `1`.  Should we just use a single code path for both ?

**Comment by gnodet** [comment]
> @slawekjaranowski the code for multi-threaded deploy should work if the number of threads is set to `1`. Should we just use a single code path for both ?

In case we really want to avoid creating a thread pool with 1 thread, we could use https://guava.dev/releases/19.0/api/docs/com/google/common/util/concurrent/MoreExecutors.html#newDirectExecutorService() ...

**Comment by slawekjaranowski** [comment]
There are two code paths:
 - number of threads <= 1 - collect all artifacts from all module and call in **one** request for Resolver
 - number of threads > 1 - create thread pool and create **separate** deploy request per module

**Comment by slawekjaranowski** [comment]
@gnodet Is my answer sufficient for you?

**Comment by gnodet** [comment]
> @gnodet Is my answer sufficient for you?

Yes, though, due to the locking that happens inside the resolver, I wonder if it would make more sense to have this parallel upload done inside the resolver rather than in the deploy plugin.  @cstamas ?

**Review by gnodet — APPROVED** [review]

**Comment by slawekjaranowski** [comment]
> Yes, though, due to the locking that happens inside the resolver, I wonder if it would make more sense to have this parallel upload done inside the resolver rather than in the deploy plugin.

That is the reason that I execute deploy for whole module in separate thread - we have the same situation like executing Maven with `-T`, each module will be deployed in own thread.

File: src/main/java/org/apache/maven/plugins/deploy/DeployMojo.java [hunk 1]
```diff
@@ -71,6 +75,14 @@ public class DeployMojo extends AbstractDeployMojo {
     @Parameter(defaultValue = "false", property = "deployAtEnd")
     private boolean deployAtEnd;
 
+    /**
+     * The number of parallel threads which will be used for deployment with the {@code deployAtEnd} feature.
+     *
+     * @since 3.1.0
+     */
+    @Parameter(property = "maven.deploy.parallelThreads", defaultValue = "1")
```
Comment (mattnelson) [inline src/main/java/org/apache/maven/plugins/deploy/DeployMojo.java:83]: sorry for the late review.

Would it be possible to support dynamic syntax like is done with the CLI parallelism flag `-T 1C` that way the deploy threads could scale with the hardware without being tightly coupled to config.

Comment (slawekjaranowski) [reply src/main/java/org/apache/maven/plugins/deploy/DeployMojo.java:83]: Can be ... 😄

**Comment by jira-importer** [comment]
Resolve #480

**Pull request merged**